File: Project_4/BT5/E_1_1.py
def read_dimacs(file_path):
    with open(file_path, 'r') as f:
        lines = f.readlines()
    
    n, m = 0, 0
    edges = []
    
    for line in lines:
        if line.startswith('c'):
            continue  # skip comment
        elif line.startswith('p'):
            _, _, n, m = line.strip().split()
            n, m = int(n), int(m)
        elif line.startswith('e'):
            _, u, v = line.strip().split()
            u, v = int(u), int(v)
            edges.append((u, v))
    
    return n, m, edges


def write_dimacs(n, edges, file_path):
    with open(file_path, 'w') as f:
        f.write(f"p edge {n} {len(edges)}\n")
        for u, v in edges:
            f.write(f"e {u} {v}\n")

File: Project_4/BT5/E_1_1.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>

using namespace std;

using Edge = pair<int, int>;

// Reads DIMACS format graph from file
bool read_dimacs(const string& file_path, int& n, int& m, vector<Edge>& edges) {
    ifstream infile(file_path);
    if (!infile.is_open()) return false;

    string line;
    edges.clear();

    while (getline(infile, line)) {
        if (line.empty() || line[0] == 'c') continue;

        istringstream iss(line);
        if (line[0] == 'p') {
            string p, format;
            iss >> p >> format >> n >> m;
        } else if (line[0] == 'e') {
            char e;
            int u, v;
            iss >> e >> u >> v;
            edges.emplace_back(u, v);
        }
    }

    infile.close();
    return true;
}

// Writes DIMACS format graph to file
bool write_dimacs(int n, const vector<Edge>& edges, const string& file_path) {
    ofstream outfile(file_path);
    if (!outfile.is_open()) return false;

    outfile << "p edge " << n << " " << edges.size() << "\n";
    for (const auto& [u, v] : edges) {
        outfile << "e " << u << " " << v << "\n";
    }

    outfile.close();
    return true;
}
